pgrx/src/fcall.rs

```diff
@@ -0,0 +1,461 @@
+    NoDefaultArguments,
+
+    #[error("Argument #{0} does not have a DEFAULT value")]
+    NotDefaultArgument(usize),
+
+    #[error("Argument's default value is not a constant expression")]
+    DefaultNotConstantExpression,
+}
+
+pub type Result<T> = std::result::Result<T, FCallError>;
+
+/// Dynamically call a named function in the current database.  The function must be one previously
+/// defined by `CREATE FUNCTION`.  Its underlying `LANGUAGE` is irrelevant -- call `LANGUAGE sql`,
+/// `LANGUAGE plpgsql`, `LANGUAGE plperl`, or (our favorite) `LANGUAGE plrust` functions.
+///
+/// The function name itself, `fname`, is an optionally schema-qualified function name identifier.
+/// If `fname` is not schema qualified, then the standard Postgres rules for searching the active
+/// `SEARCH_PATH` are followed.
+///
+/// When resolving a function by name, its argument count and types must also be considered.  These
+/// are determined by the types applied to the [`Arg`] variants of each provided argument.  To
+/// help avoid ambiguities during function resolution, even the [`Arg::Default`] and [`Arg::Null`]
+/// variants are typed -- [`Arg`] itself is generic.
+///
+/// If the called function is declared `STRICT` and at least one of the `args` are the [`Arg::Null`]
+/// variant, then the function is **not** actually called, and an `Ok(None)` is returned.
+///
+/// Arguments with `DEFAULT` clauses always appear as the last function arguments.  If you want the
+/// default value as defined by the function, either use [`Arg::Default`] or elide the default
+/// arguments entirely -- [`fcall`] will automatically fill them in.
+///
+/// # Returns
+///
+/// [`fcall`] returns an instantiated Rust type of the return value from the called function.  If
+/// that function returns NULL, [`fcall`] returns `Ok(None)`.
+///
+/// # Errors
+///
+/// Any of the [`FCallError`] variants could be returned.
+///
+/// # Panics
+///
+/// [`fcall`] itself should not panic, but it may raise assertion panics if unexpected conditions are
+/// detected.  These would indicate a bug and should be reported.
+///
+/// Note that if the function being called raises a Postgres ERROR, it is not caught by [`fcall`]
+/// and will immediately abort the transaction.  To catch such errors yourself, use [`fcall`] with
+/// [`PgTryBuilder`].
+///
+/// # Limitations
+///
+/// Currently, [`fcall`] does not support:
+///     - Functions that `RETURNS SET OF $type` or `RETURNS TABLE(...)`
+///     - Functions with `IN_OUT` or `OUT` arguments
+///     - Postgres 11
+///
+/// # Examples
+///
+/// ## Calling a UDF
+///
+/// ```sql
+/// CREATE FUNCTION sum_array(a int4[]) RETURNS int4 LANGUAGE sql AS $$ SELECT sum(value) FROM (SELECT unnest($1) value) x $$;
+/// ```
+///
+/// ```rust,no_run
+/// use pgrx::fcall::{Arg, fcall};
+///
+/// let array = vec![1,2,3];
+/// let sum = fcall::<i32>("sum_array", &[&Arg::Value(array)]);
+/// assert_eq!(sum, Ok(Some(6)));
+/// ```
```

Comment:
What's the UX like for calling functions that normally used named notation? Is that even possible with this interface, or would we have to have another? Does Postgres internally canonize them to positional notation and you have to match the positional notation anyway, even if the argument fields get swapped?

Reply:
Yeah, postgres rewrites them into positional args.  I didn't feel like it was worth supporting.  We could in the future with a different function that takes a Map instead of a slice for the argument set.